### PR TITLE
ci(*): merge build action

### PR DESCRIPTION
**Pull request by Mossaka**
this commit merges the build step in the release action and the one in the ci action since they share a large overlap code.

the build step not only builds the binary, but signs it, and upload as action artifacts. this step is now running for every ci run, including PR and release pipeline.

### CODE REVIEW

**Comment by jsturtevant** [comment]
Are there implications for signing the binary every PR?

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -85,92 +85,22 @@ jobs:
...
-      - name: Upload artifacts
-        if: ${{ matrix.is_shim == 'true' }}
-        uses: actions/upload-artifact@master
-        with:
-          name: containerd-shim-${{ matrix.runtime }}-${{ matrix.arch }}
-          path: dist/containerd-shim-${{ matrix.runtime }}-${{ matrix.arch }}.tar.gz
+    uses: ./.github/workflows/action-build.yml
+    with:
+      os: "ubuntu-22.04"
+      runtime: ${{ matrix.runtime }}
```
Comment (jsturtevant) [inline .github/workflows/release.yml:100]: there will be times when this value won't be `common` as the check in the action is expecting

**Comment by Mossaka** [comment]
> Are there implications for signing the binary every PR?

Since we are signing the artifacts that we publish on every build, I can imagine increased security and potentially degraded performance and increased complexity of managing keys. 

Although stated in the release doc for how to verify the binaries, we probably want to make that even clearer by copying the exact verification command in the release doc everytime we release signed artifacts. 

I am not sure if anyone would want to verify build artifacts for each PR / commit, but if they want, it's do-able.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -85,92 +85,22 @@ jobs:
 
   build-and-sign:
     permissions:
-      # cosign uses the GitHub OIDC token
       id-token: write
+      contents: write
```
Comment (jsturtevant) [inline .github/workflows/release.yml:89]: What was content write used for? I don't see anything that introduced that

Comment (Mossaka) [reply .github/workflows/release.yml:89]: That was added for testing purposes. Good catch! will remove that

File: .github/workflows/action-test-smoke.yml [hunk 3]
```diff
@@ -22,13 +22,13 @@ jobs:
       - name: Download artifacts
         uses: actions/download-artifact@master
         with:
-          name: containerd-shim-${{ inputs.runtime }}-linux-musl
+          name: containerd-shim-${{ inputs.runtime }}-x86_64-linux-musl
```
Comment (jsturtevant) [inline .github/workflows/action-test-smoke.yml:25]: interesting, I don't see where this arch was being included in the `inputs.runtime` before.  Or is this extra field new?

File: .github/workflows/release.yml (hunk 1 above)
Comment (Mossaka) [reply .github/workflows/release.yml:100]: Sorry, can you elaborate your concerns here?

File: .github/workflows/action-test-smoke.yml (hunk 3 above)
Comment (Mossaka) [reply .github/workflows/action-test-smoke.yml:25]: Before this PR, only the release pipeline runs both `x86_64` and `aarch64`. The ci does not have a matrix for different archs. I added that to make the build job on par with the one in the release CI.

File: .github/workflows/release.yml (hunk 1 above)
Comment (jsturtevant) [reply .github/workflows/release.yml:99]: This value in the release.yml will be something like `oci-tar-builder` https://github.com/containerd/runwasi/blob/8c98100eb74008844975aa48a6521f618f0eef50/.github/workflows/release.yml#L66-L69

and I think signing will fail in the new signing step at https://github.com/containerd/runwasi/pull/548/files#diff-43035769cde349b680ebf72f18e67b049e730d0d0ec91174e6455a685b3c2accR61-R64

**Comment by marosset** [comment]
If you add `# yaml-language-server: $schema=https://json.schemastore.org/github-action.json` to the top of all the github action files you'll get schema validation in a lot of editors like vs code

File: .github/workflows/action-test-smoke.yml [hunk 4]
```diff
@@ -22,13 +22,13 @@ jobs:
...
           path: dist
       - name: Unpack artifats
         shell: bash
         run: |
           mkdir -p dist/bin
-          tar -xzf dist/containerd-shim-${{ inputs.runtime }}-linux-musl.tar.gz -C dist/bin
+          tar -xzf dist/containerd-shim-${{ inputs.runtime }}-x86_64-linux-musl.tar.gz -C dist/bin
```
Comment (marosset) [inline .github/workflows/action-test-smoke.yml:31]: nit: Maybe make this a makefile target since the tar command is repeated in multiple workflows?

**Comment by Mossaka** [comment]
> Are there implications for signing the binary every PR?

Now that I thought more about the implications, and realized that I missed one of the biggest implication in my previous comment - signing requires feature branches to be pushed to upstream, because they require `id-token: write` permission which will not grant to public forked repos (the max permission for `id-token` is `read` for forked repos, so even though you specified `write` permission, the pipeline won't grant you the context you want). 

I am okay to put a condition on the signing to differentiate build / release pipelines so we can disable signing on build. Wdyt?

File: .github/workflows/release.yml (hunk 1 above)
Comment (Mossaka) [reply .github/workflows/release.yml:106]: Good point! I have modified the logic of `runtime_sub` step to replace the name of non-shim crate to `common`.

File: .github/workflows/action-build.yml [hunk 5]
```diff
@@ -15,13 +19,24 @@ on:
...
+  build-sign-upload:
+    permissions:
+      id-token: write
     name: build for ${{ inputs.slug }}
     runs-on: ${{ inputs.os }}
     steps:
-      - uses: actions/checkout@v3
+      - name: describe runner
+        run: |
+          echo "Running job with os: '${{ inputs.os }}', arch: '${{ inputs.arch }}', slug: '${{ inputs.slug }}', runtime: '${{ inputs.runtime }}', target: '${{ inputs.target }}'"
```
Comment (cpuguy83) [inline .github/workflows/action-build.yml:38]: Maybe you can write this such that it shows up as an annotation? https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#example-setting-a-notice-message

File: .github/workflows/action-build.yml [hunk 6]
```diff
@@ -35,12 +50,51 @@ jobs:
...
+        shell: bash
+        run: echo "OPT_PROFILE=release" >> ${GITHUB_ENV}
       - name: Build
         run: make build-${{ inputs.runtime }}
       - name: Run tests
         timeout-minutes: 5
         run: |
           make test-${{ inputs.runtime }}
+        if: ${{ inputs.arch == 'x86_64' }}
+      - name: Sign the binary
```
Comment (cpuguy83) [inline .github/workflows/action-build.yml:68]: Maybe we can move signing to a separate action that we call from here?
We can move the `Setup cosign for signing` part in there as well.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: .github/workflows/action-build.yml (hunk 5 above)
Comment (Mossaka) [reply .github/workflows/action-build.yml:38]: I am not sure if I understand the benefits of creating annotations. The example showed in the doc has a file name and a line number, which isn't applied to what we want to show here. 

Perhaps, a job summary would be better suited? GITHUB_STEP_SUMMARY

Comment (Mossaka) [reply .github/workflows/action-build.yml]: I've added `error` and `notice` annotations to logs.

File: .github/workflows/action-build.yml (hunk 6 above)
Comment (Mossaka) [reply .github/workflows/action-build.yml:63]: Done

**Comment by Mossaka** [comment]
Could you please this PR again? @cpuguy83 @devigned 🙏

**Review by devigned — APPROVED** [review]
lgtm!

Nice work, @Mossaka!

**Comment by Mossaka** [comment]
Going to merge this now and any follow up can be addressed later

**Pull request merged**